pynicotine/natpmp.py

```diff
@@ -0,0 +1,643 @@
+
+RENEWAL_INTERVAL = 14400  # 4 hours
+
+import os
+import re
+import struct
+import socket
+import select
+import platform
+
+# Modified version of NAT-PMP client library from https://github.com/yimingliu/py-natpmp
+# 
+# Provides functions to interact with NAT-PMP gateways implementing version 0
+# of the NAT-PMP draft specification.
+# 
+# This version does not completely implement the draft standard.
+# * It does not provide functionality to listen for address change packets.
+# * It does not have a proper request queuing system, meaning that
+# multiple requests may be issued in parallel, against spec recommendations.
+# 
+# For more information on NAT-PMP, see the NAT-PMP draft specification:
+# http://files.dns-sd.org/draft-cheshire-nat-pmp.txt
+
+# Copyright (c) 2008-2016, Yiming Liu, All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without modification,
+# are permitted provided that the following conditions are met:
+# 
+# * Redistributions of source code must retain the above copyright notice,
+#   this list of conditions and the following disclaimer.
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+# * The names of the author and contributors may not be used to endorse or promote
+#   products derived from this software without specific prior written permission.
+# 
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 'AS IS'
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+NATPMP_PORT = 5351
+
+NATPMP_RESERVED_VAL = 0
+
+OP_UDP = 1
+OP_TCP = 2
+
+def op_str(op):
+    if op == OP_UDP:
+        return "UDP"
+    elif op == OP_TCP:
+        return "TCP"
+
+    return None
+
+class Result:
+    SUCCESS = 0  # Success
+    UNSUPPORTED_VERSION = 1  # Unsupported Version
+    NOT_AUTHORIZED = 2  # Not Authorized/Refused/NATPMP turned off
+    NETWORK_FAILURE = 3  # Network Failure
+    OUT_OF_RESOURCES = 4  # can not create more mappings
+    UNSUPPORTED_OP = 5  # not a supported opcode
+    # all remaining results are fatal errors
+    GATEWAY_NO_VALID_GATEWAY = -10
+    GATEWAY_NO_SUPPORT = -11
+    GATEWAY_CANNOT_FIND = -12
+
+    MSG_DICT = {
```

Comment:
Please remove all the full-stops we don't use `.` at the end of any short strings.